File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    def get_total_time(level):
        total_time = times[0]  # 첫 번째 퍼즐은 무조건 한 번에 해결
        for i in range(1, len(diffs)):
            if diffs[i] <= level:
                total_time += times[i]
            else:
                retries = diffs[i] - level
                total_time += retries * (times[i] + times[i - 1]) + times[i]
        return total_time

    left, right = 1, max(diffs)
    answer = right

    while left <= right:
        mid = (left + right) // 2
        if get_total_time(mid) <= limit:
            answer = mid
            right = mid - 1
        else:
            left = mid + 1

    return answer

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.js
function solution(diffs, times, limit) {
    let left = 1
    let right = 0;
    
    diffs.forEach((item) => {
        if (right < item) right = item;
    });

    while (left <= right) {
        const mid = Math.floor((left + right) / 2)
        const result = checkTime(diffs, times, mid)
        
        if (result < limit) {
            right = mid - 1
        } else if (result > limit) {
            left = mid + 1
        } else {
            return mid
        }
    }
    return left
}

function checkTime(diffs, times, level) {
    let time = 0
    
    for (let i = 0; i < diffs.length; i++) {
        if (diffs[i] <= level) {
            time += times[i]
        } else {
            const prevTime = i > 0 ? times[i - 1] : 0;
            time += (diffs[i] - level) * (times[i] + prevTime) + times[i];

        }
    }
    
    return time
}